### PR TITLE
Adding new Controller API for updating the instance config in the cluster

### PR DESCRIPTION
**Pull request by icefury71**
This PR addresses the issue raised in #4082.
It introduces a new Controller API to update the instance config for an existing instance in the cluster.

## Release Notes
New Controller API introduced:
- Update an existing instance config: `PUT /instances/{instanceName}` with `Instance` object as the pay-load

### CODE REVIEW

**Comment by mcvsubbu** [comment]
Helix provides commands to add/remove/list tags, so the only value added in making this a part of pinot is to make sure that the tags correspond to the conventions used/allowed in pinot.  Even this is not completely predictable, because we can use pretty much any tag name in [tagOverrideConfig](https://github.com/apache/incubator-pinot/blob/master/pinot-common/src/main/java/org/apache/pinot/common/config/TagOverrideConfig.java).  I suggest that we refer the admins to the helix commands to do this operation instead of adding code in pinot. We can add documentation on specific helix commands in pinot docs.

**Comment by icefury71** [comment]
> Helix provides commands to add/remove/list tags, so the only value added in making this a part of pinot is to make sure that the tags correspond to the conventions used/allowed in pinot. Even this is not completely predictable, because we can use pretty much any tag name in [tagOverrideConfig](https://github.com/apache/incubator-pinot/blob/master/pinot-common/src/main/java/org/apache/pinot/common/config/TagOverrideConfig.java). I suggest that we refer the admins to the helix commands to do this operation instead of adding code in pinot. We can add documentation on specific helix commands in pinot docs.

Thanks for your comments @mcvsubbu . I believe this API is still useful for the following reasons:
1) Adding a "tenant" as a tag is a very Pinot specific operation. We may not be able to do pre-requisite checks if we do this through helix (eg: retagging from one tenant to another).
2) Having one admin API endpoint is very convenient and allows for the caller (eg: automated cluster/setup tenant, node replacement etc).

Lemme know if you have any more questions.

**Comment by snleee** [comment]
@mcvsubbu I agree with @icefury71. We are depending on helix's tagging feature for our tenant implementation and adding/removing tenant is Pinot specific operation. Our user should be able to control Pinot fully using admin APIs provided by us.

+1 on the fact that this API will make sure that tenant tags are in the correct form while using Helix API directly won't do any check.

**Comment by snleee** [comment]
@icefury71 Current implementation adds a generic API that can add arbitrary `tags` on `instance` (there's no check on `newTag`). I think that this is why @mcvsubbu mentioned that we are probably adding the same API that Helix already provides. We currently use Helix's tag mostly for tenant assignment but it's possible that we will use these tags for other features (e.g. multiple tags for multiple features for an instance).  In that case, this API can start to become complex since it will have to include all the validations for different tag usages.

Instead, I think that it's better to provide very specific API on `/tenants` and limit its functionality to add/remove instances to/from a tenant and not to touch other tags.

I can think of something like the following:
```
List the instances that belong to a tenant (this already exists)
GET /tenant/{tenantName}

Add an instance to a tenant
POST /tenant/{tenantName}/{instanceName}?type=offline/realtime/broker

Delete an instance from a tenant
DELETE /tenant/{tenantName}/{instanceName}
```

**Comment by mcvsubbu** [comment]
@snleee  right now we have a use case where we need to tag hosts in helix that do not conform to any "tenant" scheme that you have mentioned (the tag override scenario I mentioned). In that case, we will still need to invoke the helix API, right? Or, am I missing something?

**Comment by icefury71** [comment]
@snleee I can make that change once @mcvsubbu signs off (I do agree `/tenant` is a better endpoint).

**Comment by mcvsubbu** [comment]
@snleee  and I discussed. We need to add the ability to add/remove/list tags in order to enable override configs as well. See https://github.com/apache/incubator-pinot/blob/master/pinot-common/src/main/java/org/apache/pinot/common/config/TagOverrideConfig.java

Then we have a complete solution that does not need a user to use helix command line tools to create tags. Otherwise, our solution is still incomplete.

@kishoreg  do you have an opinion on this?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java [hunk 1]
```diff
@@ -176,4 +177,68 @@ public SuccessResponse dropInstance(
...
     return new SuccessResponse("Successfully dropped instance");
   }
-}
+
+  @GET
+  @Path("/instances/{instanceName}/tags")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Get all the tags for the specified instance", produces = MediaType.APPLICATION_JSON)
+  @ApiResponses(value = {@ApiResponse(code = 200, message = "Success"), @ApiResponse(code = 404, message = "Instance not found"), @ApiResponse(code = 500, message = "Internal error")})
+  public String getInstanceTags(
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java]: IMO `getInstance()` is good enough for this purpose, no need to add another one?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java [hunk 2]
```diff
@@ -176,4 +177,68 @@ public SuccessResponse dropInstance(
...
+   * This API is used to tag a previously untagged instance. This is used when allocating a new
+   * instance to a particular tenant. See https://github.com/apache/incubator-pinot/issues/4082
+   */
+  @POST
+  @Path("/instances/{instanceName}/tags")
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Add a tag to an untagged instance", notes = "Add a tag to an untagged instance")
+  @ApiResponses(value = {@ApiResponse(code = 200, message = "Success"), @ApiResponse(code = 400, message = "Bad Request"), @ApiResponse(code = 404, message = "Instance not found"), @ApiResponse(code = 409, message = "Instance cannot be dropped"), @ApiResponse(code = 500, message = "Internal error")})
+  public SuccessResponse addInstanceTag(
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java]: I suggest adding a `PUT /instances/{instanceName}` to modify the existing instance configs (similar to the existing `POST /instances` which adds a new instance config), that will be much more general than this

**Comment by snleee** [comment]
@mcvsubbu @icefury71 
I'm actually convinced that adding new APIs under`/tenant` also may not be a good option. (@mcvsubbu pointed out that the existing tenant concept cannot handle `tagOverrideConfigs`. We need to add more flexible API that can add/remove any arbitrarily named tags for the full support of `tagOverrideConfig` feature).

In that case, I agree with Jackie's idea to add `PUT /instances/{instanceName}`. With this API, user have a full flexibility to manipulate the tags for instances.

**Comment by icefury71** [comment]
> @mcvsubbu @icefury71
> I'm actually convinced that adding new APIs under`/tenant` also may not be a good option. (@mcvsubbu pointed out that the existing tenant concept cannot handle `tagOverrideConfigs`. We need to add more flexible API that can add/remove any arbitrarily named tags for the full support of `tagOverrideConfig` feature).
> 
> In that case, I agree with Jackie's idea to add `PUT /instances/{instanceName}`. With this API, user have the full flexibility to manipulate tags on an instance.

Makes sense. I can make that change

**Comment by snleee** [comment]
@mcvsubbu and I discussed offline and we agreed that `PUT /instances/{instanceName}` seems to be the best approach to take. 

@icefury71 Can you make the change to API?

**Comment by mcvsubbu** [comment]
I am fine adding a PUT on the instance config to modify the tag list. We have to have a caveat that PUT cannot be executed in multiple threads. For example, if someone runs PUT to add a tag A, and another runs a PUT to add tag B, only one of these may be added, and the other one ignored.

In order to make things better, we should probably return the znode version in GET, and accept ifNotMofied Header on the PUT operation.

**Comment by icefury71** [comment]
> I am fine adding a PUT on the instance config to modify the tag list. We have to have a caveat that PUT cannot be executed in multiple threads. For example, if someone runs PUT to add a tag A, and another runs a PUT to add tag B, only one of these may be added, and the other one ignored.
> 
> In order to make things better, we should probably return the znode version in GET, and accept ifNotMofied Header on the PUT operation.

@mcvsubbu modified as per suggestion. The PUT API is internally calling a synchronized update method which should take care of the concurrency concern.

**Comment by mcvsubbu** [comment]
Please update the commit message.
Also, i am not sure where (or even how) to document the fact that if multiple updates happen, not all of them will be taken. For now, we can add that as a part of the commit message at least, I suppose. thanks.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java [hunk 3]
```diff
@@ -176,4 +178,20 @@ public SuccessResponse dropInstance(
...
+  public SuccessResponse updateInstance(
+      @ApiParam(value = "Instance name", required = true, example = "Server_a.b.com_20000 | Broker_my.broker.com_30000") @PathParam("instanceName") String instanceName,
+      Instance instance) {
+    LOGGER.info("Instance update request received for instance: {}", instanceName);
+    if (!pinotHelixResourceManager.updateInstance(instanceName, instance).isSuccessful()) {
+      throw new ControllerApplicationException(LOGGER, "Failure to update instance", Response.Status.INTERNAL_SERVER_ERROR);
+    }
+    return new SuccessResponse("Instance successfully updated");
+  }
+}
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java]: (nit) new line

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 4]
```diff
@@ -358,6 +358,19 @@ public synchronized PinotResourceManagerResponse addInstance(Instance instance)
...
 
+  /**
+   * Update a given instance for the specified Instance ID
+   */
+  public synchronized PinotResourceManagerResponse updateInstance(String instanceIdToUpdate, Instance newInstance) {
+    List<String> instances = getAllInstances();
+    if (!instances.contains(instanceIdToUpdate)) {
+      return PinotResourceManagerResponse.failure("Instance " + instanceIdToUpdate + " does not exists");
+    } else {
+      _helixAdmin.setInstanceConfig(_helixClusterName, instanceIdToUpdate, newInstance.toInstanceConfig());
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java]: This won't be able the change the instance host/port. In order to do that, you need to use data accessor to set it.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 5]
```diff
@@ -361,6 +361,23 @@ public synchronized PinotResourceManagerResponse addInstance(Instance instance)
...
+   * Update a given instance for the specified Instance ID
+   */
+  public synchronized PinotResourceManagerResponse updateInstance(String instanceIdToUpdate, Instance newInstance) {
+    List<String> instances = getAllInstances();
+    if (!instances.contains(instanceIdToUpdate)) {
+      return PinotResourceManagerResponse.failure("Instance " + instanceIdToUpdate + " does not exists");
+    } else {
+      InstanceConfig newConfig = InstanceUtils.toHelixInstanceConfig(newInstance);
+      if(!_helixZkManager.getHelixDataAccessor().setProperty(
+          _helixZkManager.getHelixDataAccessor().keyBuilder().instanceConfig(instanceIdToUpdate), newConfig)) {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:374]: For performance and simplicity:
```suggestion
    if (getHelixInstanceConfig(instanceIdToUpdate) == null) {
      return PinotResourceManagerResponse.failure("Instance " + instanceIdToUpdate + " does not exists");
    } else {
      InstanceConfig newConfig = InstanceUtils.toHelixInstanceConfig(newInstance);
      if(!_helixDataAccessor.setProperty(_keyBuilder.instanceConfig(instanceIdToUpdate), newConfig)) {
```

**Pull request merged**